torch/_inductor/codegen/common.py

```diff
@@ -135,48 +130,68 @@ def deduce_node_dtype_by_inputs(self, node: torch.fx.Node):
             [n.meta[OptimizationContext.key].dtype for n in input_nodes],
         )
 
+    def deduce_node_dtype_by_subgraph(self, node: torch.fx.Node):
+        sub_graph = self.graphs[node.target]
+        return self.propagate_graph(sub_graph)
+
     def deduce_node_dtype(self, node: torch.fx.Node):
         if node.target in OpDtypeClassifier.boolean_ops():
             return torch.bool
 
-    if new_node_propogated:
-        _data_type_propagation(sub_graph, masked_subblock_dtypes)
+        if node.op in OpDtypeClassifier.placeholder_ops():
+            return None
 
         if node.target in OpDtypeClassifier.explicit_dtype_ops():
             return node.args[-1]
 
         if node.target in OpDtypeClassifier.rand_ops():
             return torch.float
 
-    assert isinstance(node, SchedulerNode)
-    _node: SchedulerNode = node
+        if node.target in OpDtypeClassifier.index_ops():
+            return torch.int64
+
+        if node.target in OpDtypeClassifier.load_store_ops():
+            buf_name = node.args[1]
+            return V.graph.get_dtype(buf_name)
+
+        if node.target in OpDtypeClassifier.reduction_ops():
+            _, _, dtype, _, _, _, _ = node.args
+            return dtype
 
-    def get_masked_subblock_data_type(graph: torch.fx.Graph):
-        output_dtype = None
+        if any(
+            node.target.startswith(op) for op in OpDtypeClassifier.with_subblock_ops()
+        ):
+            return self.deduce_node_dtype_by_subgraph(node)
+
+        return self.deduce_node_dtype_by_inputs(node)
 
+    def propagate_graph(self, graph: torch.fx.Graph):
         for node in graph.nodes:
-            if node.target == "output":
-                assert output_dtype is None
+            if OptimizationContext.key in node.meta:
                 opt_ctx = node.meta[OptimizationContext.key]
-                assert opt_ctx
-                assert opt_ctx.dtype
-                output_dtype = opt_ctx.dtype
-
-        return output_dtype
-
-    if isinstance(_node._body, LoopBody):
-        body: LoopBody = node._body
-        root_block = body.root_block
-        sub_blocks = body.subblocks
-        masked_subblock_dtypes = {}
-        for sub_block_name, sub_block in sub_blocks.items():
-            assert "masked_subblock" in sub_block_name
-            _sub_graph: torch.fx.Graph = sub_block.graph
-            _data_type_propagation(_sub_graph, masked_subblock_dtypes)
-            masked_subblock_dtype = get_masked_subblock_data_type(_sub_graph)
-            masked_subblock_dtypes[sub_block_name] = masked_subblock_dtype
-
-        _data_type_propagation(root_block.graph, masked_subblock_dtypes)
+            else:
+                opt_ctx = OptimizationContext()
+
+            opt_ctx.dtype = self.deduce_node_dtype(node)
+            node.meta[OptimizationContext.key] = opt_ctx
+        if node.target == "output":
+            return node.meta[OptimizationContext.key].dtype
```

Comment:
nit: not all path returns value. Also, please add an assertion on `graph.nodes` is not empty. `node` would be undefined if its empty.

Reply:
Thanks for advice, added assertion on `graph.nodes` and also add assertion on `return dtype for subblock case`

Reply:
You also need to make sure all paths return value. Also, is it the right assumption that `output` node is always the last one?